### DISCUSSION TITLE
Health economics model specific in-code checks i.e. asserts

### DISCUSSION THREAD

**Discussion by n8thangreen** [HTAhackathon2024/.github#8] [Ideas]
### What is the problem we're trying to solve

Checking that the code does what it should at run time is obviously done by unit testing but these test function outputs. Asserts are checks inside of the code which check model inputs or derived objects inside functions.

HE models has common things like checking that a transition probability matrix is correct.

### How will we tackle the problem?

This is something that has been started with the [assertHE](https://github.com/dark-peak-analytics/assertHE) package, but its functionality is moving away from asserts and towards other features.

The idea is to
- add to the available asserts
- extend feature like handling exceptions more elegantly e.g. [tryCatch](https://medium.com/number-around-us/catch-me-if-you-can-exception-handling-in-r-2e0f6c473a28)
- including a [logger](https://cran.r-project.org/web/packages/logger/vignettes/Intro.html) so that the result of the asserts are passed to file

### Which tools will we use?

There are various logger packages, e.g. `{logger}`, `{logging}` or `{futile.logger}`

### Previous work

[assertHE](https://github.com/dark-peak-analytics/assertHE)

### Any further resources

* https://blog.djnavarro.net/posts/2023-08-08_being-assertive/
* [asserthat](https://github.com/hadley/assertthat)

**Comment by RobertASmith** [comment]
Hi Nathan and everyone,

I thought it may be useful to provide more context about the assertHE package. A talk can be found [here](https://www.youtube.com/watch?v=wr8-w-6QGno). 

There are two main uses of the package. 
1) The asserts are a very simple but high value-add set of functions which can be used by everybody to check code confirms to a set of rules. These are the easiest for people to get involved in. A logger would be cool and maybe be something I'd want to work on while helping out.
2) The visualiser is cool, but harder to get involved in. There are a number of different things going on there - see the [Issues](https://github.com/dark-peak-analytics/assertHE/issues) tab of the repo to see more work in progress including adding external package dependencies, an RStudio plug-in and adapting the LLM integration method.

Many thanks for your interest. 

Rob

**Comment by oliverodolin** [comment]
Would love to help out with this at the hackathon! I'd like to get better at writing asserts and become more familiar with the R packages for doing this. Would also be cool to learn more about exception handling. That being said -- I don't have experience building HE models, so if the assertions require advanced technical knowledge, then I won't be able to help much.

**Comment by RobertASmith** [comment]
Early draft manuscript still being prepared:

[assertHE_manuscript.pdf](https://github.com/user-attachments/files/16454015/assertHE_manuscript.pdf)

**Comment by carolinaheu** [comment]
I think this is an interesting idea also. I have not had time to read the papers Robert mentions. In contrast to Oliver, I have more economics technical knowledge and less R technical knowledge, so it could be a good addition to the team. I also have some experience with technical validation of models which could be of help?

**Reply by RobertASmith** [reply]
This sounds great. I think combining the HE specific knowledge about the rules that different outputs (e.g. transition probabilities or survival extrapolations) should have to comply to (@carolinaheu), with interest in developing these checks in R (@oliverodolin) should give us a useful set of outputs.

**Comment by Neil-W1** [comment]
I would be interested in joining this too.
